caffe2/core/dispatch/DispatchTable.h

```diff
@@ -1,104 +1,118 @@
 #pragma once
 
-#include "caffe2/utils/flat_hash_map/flat_hash_map.h"
-#include "caffe2/utils/Metaprogramming.h"
 #include "caffe2/core/dispatch/OpSchema.h"
+#include "caffe2/utils/Metaprogramming.h"
+#include "caffe2/utils/flat_hash_map/flat_hash_map.h"
 
-#include <type_traits>
 #include <array>
-#include <unordered_map>
+#include <atomic>
 #include <iostream>
 #include <mutex>
+#include <type_traits>
+#include <unordered_map>
 
 namespace c10 {
 
 namespace details {
-
 /// Kernel implementations in a thread-safe hash table.
-template<class Key>
+template <class Key>
 class ThreadsafeOperatorTable_ final {
-public:
-    // TODO The current implementation below does not have the correct correctness characteristics
-    // which we need.  It's worth spelling out exactly what we need:
-    //
-    //  - We need LOCK FREE read access to the table (as per the performance benchmark
-    //    at https://fb.quip.com/hvz3AGnx8MQ8
-    //
-    //  - We need to support writes which are possibly concurrent with reads, occurring when
-    //    a dynamic library is loaded or unloaded.
-    //
-    //  - We probably can require that dynamic library loads/unloads be synchronized (so
-    //    there are never two concurrent loads.)
-
-    template<class Key_>
-    void emplace(Key_&& key, void* value) {
-      using std::to_string;
-      // TODO Locking
-      //std::unique_lock<std::shared_timed_mutex> lock(mutex_);
-
-      auto result = map_.emplace(std::forward<Key>(key), value);
-      if (!result.second) {
-        std::ostringstream msg;
-        msg << "Tried to register conflicting kernels to the dispatcher: " << key;
-        throw std::logic_error(msg.str());
-      }
+ public:
+  ThreadsafeOperatorTable_() {
+    map_.store(new ska::flat_hash_map<Key, void*>());
+  }
+
+  virtual ~ThreadsafeOperatorTable_() {
+    delete map_.load();
+    for (auto* map : maps_to_delete_) {
+      delete map;
     }
+  }
 
-    void erase(const Key& key) {
-      // TODO Locking
-      //std::unique_lock<std::shared_timed_mutex> lock(mutex_);
+  template <class Key_>
+  void emplace(Key_&& key, void* value) {
```

Comment:
Shouldn't we encourage bulk inserts? Otherwise, O(n) emplaces will cost O(n^2).

Reply:
good suggestion, but I think this is outside of the scope for this diff

Reply:
How does it cost `O(n^2)`? The hashmap will only reallocate when crossing a `2^n` size boundary, so inserting should be amortized `O(n)`.

Reply:
The current implementation is overly expensive. We should avoid allocating a new hash map every time, then we'd be amortized `O(n)`. It should be possible to just keep two hash maps around, which are both always identical. One is read from, writing writes to the other one, swaps them, then writes the same change to the first one.

Reply:
Well, it didn't use to be O(n^2), because we weren't copying the hashmap before?